builtin/credential/aws/backend.go

```diff
@@ -0,0 +1,165 @@
+)
+
+func Factory(conf *logical.BackendConfig) (logical.Backend, error) {
+	b, err := Backend(conf)
+	if err != nil {
+		return nil, err
+	}
+	return b.Setup(conf)
+}
+
+type backend struct {
+	*framework.Backend
+	Salt *salt.Salt
+
+	// Lock to make changes to any of the backend's configuration endpoints.
+	configMutex sync.RWMutex
+
+	// Duration after which the periodic function of the backend needs to
+	// tidy the blacklist and whitelist entries.
+	tidyCooldownPeriod time.Duration
+
+	// nextTidyTime holds the time at which the periodic func should initiatite
+	// the tidy operations. This is set by the periodicFunc based on the value
+	// of tidyCooldownPeriod.
+	nextTidyTime time.Time
+
+	// Map to hold the EC2 client objects indexed by region. This avoids the
+	// overhead of creating a client object for every login request. When
+	// the credentials are modified or deleted, all the cached client objects
+	// will be flushed.
+	EC2ClientsMap map[string]*ec2.EC2
+}
+
+func Backend(conf *logical.BackendConfig) (*framework.Backend, error) {
+	salt, err := salt.NewSalt(conf.StorageView, &salt.Config{
+		HashFunc: salt.SHA256Hash,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	b := &backend{
+		// Setting the periodic func to be run once in an hour.
+		// If there is a real need, this can be made configurable.
+		tidyCooldownPeriod: time.Hour,
+		Salt:               salt,
+		EC2ClientsMap:      make(map[string]*ec2.EC2),
+	}
+
+	b.Backend = &framework.Backend{
+		PeriodicFunc: b.periodicFunc,
+		AuthRenew:    b.pathLoginRenew,
+		Help:         backendHelp,
+		PathsSpecial: &logical.Paths{
+			Unauthenticated: []string{
+				"login",
+			},
+		},
+		Paths: []*framework.Path{
+			pathLogin(b),
+			pathRole(b),
+			pathListRoles(b),
+			pathRoleTag(b),
+			pathConfigClient(b),
+			pathConfigCertificate(b),
+			pathConfigTidyRoleTags(b),
+			pathConfigTidyIdentities(b),
+			pathListCertificates(b),
+			pathBlacklistRoleTag(b),
+			pathListBlacklistRoleTags(b),
+			pathTidyRoleTags(b),
+			pathWhitelistIdentity(b),
+			pathTidyIdentities(b),
+			pathListWhitelistIdentities(b),
+		},
+	}
+
+	return b.Backend, nil
+}
+
+// periodicFunc performs the tasks that the backend wishes to do periodically.
+// Currently this will be triggered once in a minute by the RollbackManager.
+//
+// The tasks being done currently by this function are to cleanup the expired
+// entries of both blacklist role tags and whitelist identities. Tidying is done
+// not once in a minute, but once in an hour, controlled by 'tidyCooldownPeriod'.
+// Tidying of blacklist and whitelist are by default enabled. This can be
+// changed using `config/tidy/roletags` and `config/tidy/identities` endpoints.
+func (b *backend) periodicFunc(req *logical.Request) error {
+
+	// Run the tidy operations for the first time. Then run it when current
+	// time matches the nextTidyTime.
+	if b.nextTidyTime.IsZero() || !time.Now().UTC().Before(b.nextTidyTime) {
```

Comment:
I think this should be persisted. Otherwise if we've just taken the hit to run it, and we do a quick Vault upgrade, it's going to run again when we start up (and every time we start up). Probably not ideal for a possibly expensive operation.